packages/cli/lib/charm.ts

```diff
@@ -217,35 +254,51 @@ export async function linkCharms(
 ): Promise<void> {
   const manager = await loadManager(config);
 
-  const sourceCharm = await manager.get(sourceCharmId, false);
-  if (!sourceCharm) {
-    throw new Error(`Source charm "${sourceCharmId}" not found`);
-  }
+  // Get source cell (charm or arbitrary cell)
+  const { cell: sourceCell, isCharm: sourceIsCharm } = await getCellByIdOrCharm(
+    manager,
+    sourceCharmId,
+    "Source",
+  );
 
-  const targetCharm = await manager.get(targetCharmId, false);
-  if (!targetCharm) {
-    throw new Error(`Target charm "${targetCharmId}" not found`);
-  }
+  // Get target cell (charm or arbitrary cell)
+  const { cell: targetCell, isCharm: targetIsCharm } = await getCellByIdOrCharm(
+    manager,
+    targetCharmId,
+    "Target",
+  );
 
   // Navigate to the source path
-  let sourceCell: Cell<any> = sourceCharm;
+  let sourceResultCell = sourceCell;
+  // For charms, manager.get() already returns the result cell, so no need to add "result"
+
   for (const segment of sourcePath) {
-    sourceCell = sourceCell.key(segment);
+    sourceResultCell = sourceResultCell.key(segment);
   }
-  const sourceCellLink = sourceCell.getAsCellLink();
 
-  // Navigate to the parent of the target path
-  let targetCell: Cell<any> = targetCharm;
+  const sourceCellLink = sourceResultCell.getAsCellLink();
+
+  // Navigate to the target path
   const targetKey = targetPath.pop();
```

Comment:
you could skip this and below just call targetInputCell.set() directly. Or was there a specific reason for this?

(it's also possible to change the top-level input of a charm, so even the non-empty constraint is technically not necessary)